ArmPkg/Drivers/ArmGicDxe/GicV5/ArmGicV5Dxe.c

```diff
@@ -0,0 +1,582 @@
+    // Only PPI and SPI are currently supported
+    DEBUG ((DEBUG_ERROR, "Unsupported GIC interrupt type\n"));
+    GicV5EndOfInterrupt (&gHardwareInterruptV5Protocol, GicInterrupt);
+    return;
+  }
+
+  InterruptHandler = mRegisteredInterruptHandlers[IntId];
+  if (InterruptHandler != NULL) {
+    // Call the registered interrupt handler.
+    InterruptHandler (GicInterrupt, SystemContext);
+  } else {
+    DEBUG ((DEBUG_ERROR, "Spurious GIC interrupt: 0x%x\n", (UINT32)IntId));
+    GicV5EndOfInterrupt (&gHardwareInterruptV5Protocol, GicInterrupt);
+  }
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+GicV5RegisterInterruptSource (
+  IN EFI_HARDWARE_INTERRUPT_PROTOCOL  *This,
+  IN HARDWARE_INTERRUPT_SOURCE        Source,
+  IN HARDWARE_INTERRUPT_HANDLER       Handler
+  )
+{
+  // The GICv5 driver supports both PPIs and SPIs. To support both in a single
+  // interrupt handler array, map IDs 0 to NUM_PPI_INTERRUPTS-1 as PPIs, and
+  // NUM_PPI_INTERRUPTS and above as SPIs.
+  UINTN  Type = Source & GICV5_INTERRUPT_TYPE_MASK;
+  UINTN  IntId;
+
+  if (Type == GICV5_INTERRUPT_TYPE_PPI) {
+    // PPI
+    IntId = Source & GICV5_INTERRUPT_ID_MASK;
+    if (IntId >= GICV5_NUM_PPI_INTERRUPTS) {
+      // GICv5 defines valid PPI IDs as 0-127
+      ASSERT (FALSE);
+      return EFI_UNSUPPORTED;
+    }
+  } else if (Type == GICV5_INTERRUPT_TYPE_SPI) {
+    // SPI
+    IntId = (Source & GICV5_INTERRUPT_ID_MASK) + GICV5_NUM_PPI_INTERRUPTS;
+  }
+
+  if (IntId >= mGicNumInterrupts) {
+    ASSERT (FALSE);
+    return EFI_UNSUPPORTED;
+  }
+
+  return GicCommonRegisterInterruptSource (
+           This,
+           Source,
+           Handler,
+           &mRegisteredInterruptHandlers[IntId]
+           );
+}
+
+// The protocol instance produced by this driver
+EFI_HARDWARE_INTERRUPT_PROTOCOL  gHardwareInterruptV5Protocol = {
+  GicV5RegisterInterruptSource,
+  GicV5EnableInterruptSource,
+  GicV5DisableInterruptSource,
+  GicV5GetInterruptSourceState,
+  GicV5EndOfInterrupt
+};
+
+/**
+  Get interrupt trigger type of an interrupt
+
+  @param This          Instance pointer for this protocol
+  @param Source        Hardware source of the interrupt.
+  @param TriggerType   Returns interrupt trigger type.
+
+  @retval EFI_SUCCESS       Source interrupt supported.
+  @retval EFI_UNSUPPORTED   Source interrupt is not supported.
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+GicV5GetTriggerType (
+  IN  EFI_HARDWARE_INTERRUPT2_PROTOCOL      *This,
+  IN  HARDWARE_INTERRUPT_SOURCE             Source,
+  OUT EFI_HARDWARE_INTERRUPT2_TRIGGER_TYPE  *TriggerType
+  )
+{
+  UINT32  InterruptType = Source & GICV5_INTERRUPT_TYPE_MASK;
+  UINT32  InterruptId   = Source & GICV5_INTERRUPT_ID_MASK;
+  UINT32  GicTriggerType;
+
+  if (InterruptType == GICV5_INTERRUPT_TYPE_SPI) {
+    if (!WaitForIrsSpiIdle ()) {
+      ASSERT (FALSE);
+      return EFI_UNSUPPORTED;
+    }
+
+    MmioWrite32 (mGicIrsConfigFrameBase + GICV5_IRS_SPI_SELR, InterruptId);
+    if (!WaitForIrsSpiIdle ()) {

```

Comment:
Can we combine these configuration MMIO accesses and preceding/subsequent waits for idle into helper functions?